### ISSUE TITLE
Upsert not working with autoincrement field

### ISSUE THREAD

**Issue by Duvel** [ujjwalguptaofficial/JsStore#104]
I was testing with the upsert functionality with an table with an autoincrement key. I noticed that instead of updating the object in the store, it added a new one.

I looked at the code of the merge request and the [specs](https://developer.mozilla.org/en-US/docs/Web/API/IDBObjectStore/put) for the put functionality and I saw an issue there:

key Optional: 
The primary key of the record you want to update (e.g. from IDBCursor.primaryKey). This is only needed for object stores that have an autoIncrement primary key, therefore the key is not in a field on the record object. In such cases, calling put(item) will always insert a new record, because it doesn't know what existing record you might want to modify.

So I'm guessing that the upsert isn't working yet for an autoincrement key.

**Comment by ujjwalguptaofficial** [comment]
Have you tried inserting the key ?  Its not working for me :(.

I got a jsfiddle link - http://jsfiddle.net/ashustariov/Wk7VF/28/. I tried everything & nothing worked. Check for yourself and let me know.

**Comment by Duvel** [comment]
I have found out why the fiddle wasn't working, apparently there is a difference between in-line and out-of-line keys: 
https://stackoverflow.com/questions/36016862/indexeddb-dataerror-data-provided-to-an-operation-does-not-meet-requirements

While playing with the fiddle I noticed that put should work if you add the id, so I'm going to try to find out why it is inserting new values instead of updatin.

**Comment by Duvel** [comment]
I have no clue why it is inserting new records, although I'm adding the correct key in the object. Any ideas how I could debug this?

**Comment by Duvel** [comment]
I tried some debugging and it seems that before the put is called, the record has gotten a new id.

**Comment by ujjwalguptaofficial** [comment]
i tried everything too, but nothing worked for me. So no clues how it will work, also there is not much info about this anywhere :(.

Let me know if you got something.

**Comment by Duvel** [comment]
I see that the fiddle is not saving, the trick is omitting the keyPath when creating the objectStore:

``` js
function createMyStore(db) {
    db.createObjectStore(storeName, {
        autoIncrement : true
    });
}

request.onsuccess = function(event) {
    log('Indexed db opened!');
    var db = event.target.result;
    var t = db.transaction([storeName], TransactionMode.RW);
    var store = t.objectStore(storeName);
    try {
        var request = store.put({
            id : 1,
            myValue : 'Hello, World!'
        }, 1);
            
        request.onsuccess = function(event) {
            log('My record saved :)!');
        };
        
        request.onerror = function() {
            log('My record was not saved :(!');
        };
        
        store.getAll().onsuccess = function(event) {
            log(JSON.stringify(event.target.result));
        };
    } catch (error) {
        log(error.name + '. ' + error.message);
    }

};
```

**Comment by Duvel** [comment]
I think I have found the error. If you have a column with both a primary key and a autoincrement, JsStore handles the autoincrement, not the IndexedDb. The autoincrement seems to be handled in the value_checker and seems to be always updated, although it is a upsert.

If I set skipDataCheck to true, the record will get updated instead of inserted again.

**Comment by ujjwalguptaofficial** [comment]
aha right, that makes sense. Thanks

**Comment by ujjwalguptaofficial** [comment]
> I see that the fiddle is not saving, the trick is omitting the keyPath when creating the objectStore:
> 
> ```js
> function createMyStore(db) {
>     db.createObjectStore(storeName, {
>         autoIncrement : true
>     });
> }
> 
> request.onsuccess = function(event) {
>     log('Indexed db opened!');
>     var db = event.target.result;
>     var t = db.transaction([storeName], TransactionMode.RW);
>     var store = t.objectStore(storeName);
>     try {
>         var request = store.put({
>             id : 1,
>             myValue : 'Hello, World!'
>         }, 1);
>             
>         request.onsuccess = function(event) {
>             log('My record saved :)!');
>         };
>         
>         request.onerror = function() {
>             log('My record was not saved :(!');
>         };
>         
>         store.getAll().onsuccess = function(event) {
>             log(JSON.stringify(event.target.result));
>         };
>     } catch (error) {
>         log(error.name + '. ' + error.message);
>     }
> 
> };
> ```

so, we cant have both primary key & autoIncrement at same time right ?

**Comment by Duvel** [comment]
> so, we can't have both primary key & autoIncrement at same time right ?

Not if you want to have control over the name of the primary key.

**Comment by ujjwalguptaofficial** [comment]
I think - an option `idbAutoIncrement` : `boolean` will work. By default it will be false & when true, jsstore will supply this to indexeddb for generating autoIncrment while creating database. This way we can also check that primaryKey along with idbAutoIncrement should not be supplied.

what do you think about this ?

**Comment by Duvel** [comment]
That sounds like a good option, I like it. You can check for it in the datacheck as well and also if the name of the id is really id and not a self defined one.

**Comment by ujjwalguptaofficial** [comment]
Actully problem was in autoincrement generation of jsstore, it was generating autoincrement value even if there was data supplied in the column. So the fix was to not generate the value when data is supplied.

This has been fixed in v 3.0.0.
